File: apps/cli/internal/selfupdate/selfupdate.go
package selfupdate

import (
	"archive/tar"
	"compress/gzip"
	"context"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"io"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"

	"yishan/apps/cli/internal/buildinfo"
)

const (
	repo       = "yishan-io/yishan-mono"
	project    = "yishan-cli"
	binaryName = "yishan"
)

// Release holds metadata for a GitHub release.
type Release struct {
	Version     string
	ArchiveURL  string
	ChecksumURL string
}

// LatestRelease resolves the latest CLI release version from GitHub.
func LatestRelease(ctx context.Context) (*Release, error) {
	// GitHub redirects /releases/latest to /releases/tag/<tag>.
	// We follow the redirect to extract the tag.
	url := fmt.Sprintf("https://github.com/%s/releases/latest", repo)

	req, err := http.NewRequestWithContext(ctx, http.MethodHead, url, nil)
	if err != nil {
		return nil, fmt.Errorf("creating request: %w", err)
	}

	// Don't follow redirects — we want the Location header.
	client := &http.Client{
		CheckRedirect: func(_ *http.Request, _ []*http.Request) error {
			return http.ErrUseLastResponse
		},
	}

	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("resolving latest release: %w", err)
	}
	defer resp.Body.Close()

	location := resp.Header.Get("Location")
	if location == "" {
		return nil, fmt.Errorf("no redirect from GitHub releases/latest (status %d)", resp.StatusCode)
	}

	// Location: https://github.com/yishan-io/yishan-mono/releases/tag/v0.12.0
	parts := strings.Split(location, "/tag/v")
	if len(parts) != 2 || parts[1] == "" {
		return nil, fmt.Errorf("unexpected release tag format in redirect: %s", location)
	}

	return releaseForVersion(parts[1]), nil
}

// ReleaseForVersion builds a Release for a specific version string.
func ReleaseForVersion(version string) *Release {
	return releaseForVersion(version)
}

func releaseForVersion(version string) *Release {
	os := normalizeOS(runtime.GOOS)
	arch := normalizeArch(runtime.GOARCH)
	archive := fmt.Sprintf("%s_%s_%s_%s.tar.gz", project, version, os, arch)
	base := fmt.Sprintf("https://github.com/%s/releases/download/v%s", repo, version)

	return &Release{
		Version:     version,
		ArchiveURL:  fmt.Sprintf("%s/%s", base, archive),
		ChecksumURL: fmt.Sprintf("%s/checksums.txt", base),
	}
}

// IsNewer returns true when the release is newer than the running binary.
func (r *Release) IsNewer() bool {
	return r.Version != buildinfo.Version && buildinfo.Version != "dev"
}

// Apply downloads, verifies, and replaces the current binary.
// If the binary was installed via Homebrew, it delegates to `brew upgrade`.
func (r *Release) Apply(ctx context.Context, progress func(string)) error {
	execPath, err := os.Executable()
	if err != nil {
		return fmt.Errorf("locating current binary: %w", err)
	}
	execPath, err = filepath.EvalSymlinks(execPath)
	if err != nil {
		return fmt.Errorf("resolving symlinks: %w", err)
	}

	// Detect Homebrew-managed install.
	if isHomebrewManaged(execPath) {
		return applyViaHomebrew(ctx, progress)
	}

	tmpDir, err := os.MkdirTemp("", "yishan-update-*")
	if err != nil {
		return fmt.Errorf("creating temp dir: %w", err)
	}
	defer os.RemoveAll(tmpDir)

	// Download archive.
	progress(fmt.Sprintf("Downloading %s...", filepath.Base(r.ArchiveURL)))
	archivePath := filepath.Join(tmpDir, "archive.tar.gz")
	if err := download(ctx, r.ArchiveURL, archivePath); err != nil {
		return fmt.Errorf("downloading archive: %w", err)
	}

	// Download checksums.
	progress("Verifying checksum...")
	checksumPath := filepath.Join(tmpDir, "checksums.txt")
	if err := download(ctx, r.ChecksumURL, checksumPath); err != nil {
		return fmt.Errorf("downloading checksums: %w", err)
	}

	// Verify checksum.
	if err := verifyChecksum(archivePath, checksumPath); err != nil {
		return fmt.Errorf("checksum verification failed: %w", err)
	}

	// Extract binary.
	progress("Extracting...")
	newBinaryPath := filepath.Join(tmpDir, binaryName)
	if err := extractBinary(archivePath, newBinaryPath); err != nil {
		return fmt.Errorf("extracting binary: %w", err)
	}

	// Replace current binary atomically.
	progress("Installing...")
	if err := replaceBinary(execPath, newBinaryPath); err != nil {
		return fmt.Errorf("replacing binary: %w", err)
	}

	return nil
}

func download(ctx context.Context, url, dest string) error {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return err
	}

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("HTTP %d from %s", resp.StatusCode, url)
	}

	f, err := os.Create(dest)
	if err != nil {
		return err
	}
	defer f.Close()

	_, err = io.Copy(f, resp.Body)
	return err
}

func verifyChecksum(archivePath, checksumPath string) error {
	checksumData, err := os.ReadFile(checksumPath)
	if err != nil {
		return fmt.Errorf("reading checksums: %w", err)
	}

	// checksums.txt contains the original archive name, not "archive.tar.gz".
	// We need to find the entry matching our OS/arch.
	osName := normalizeOS(runtime.GOOS)
	archName := normalizeArch(runtime.GOARCH)

	var expectedHash string
	for _, line := range strings.Split(string(checksumData), "\n") {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		fields := strings.Fields(line)
		if len(fields) != 2 {
			continue
		}
		// Match by OS_ARCH suffix in the filename.
		if strings.Contains(fields[1], fmt.Sprintf("%s_%s", osName, archName)) {
			expectedHash = fields[0]
			break
		}
	}

	if expectedHash == "" {
		return fmt.Errorf("no checksum found for %s/%s in checksums.txt", osName, archName)
	}

	// Compute actual hash.
	f, err := os.Open(archivePath)
	if err != nil {
		return err
	}
	defer f.Close()

	h := sha256.New()
	if _, err := io.Copy(h, f); err != nil {
		return err
	}

	actualHash := hex.EncodeToString(h.Sum(nil))
	if actualHash != expectedHash {
		return fmt.Errorf("checksum mismatch: expected %s, got %s", expectedHash, actualHash)
	}

	return nil
}

func extractBinary(archivePath, destPath string) error {
	f, err := os.Open(archivePath)
	if err != nil {
		return err
	}
	defer f.Close()

	gz, err := gzip.NewReader(f)
	if err != nil {
		return err
	}
	defer gz.Close()

	tr := tar.NewReader(gz)
	for {
		hdr, err := tr.Next()
		if err == io.EOF {
			break
		}
		if err != nil {
			return err
		}

		name := filepath.Base(hdr.Name)
		if name == binaryName || name == binaryName+".exe" {
			out, err := os.OpenFile(destPath, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0755)
			if err != nil {
				return err
			}
			defer out.Close()

			if _, err := io.Copy(out, tr); err != nil {
				return err
			}
			return nil
		}
	}

	return fmt.Errorf("binary %q not found in archive", binaryName)
}

// replaceBinary atomically replaces oldPath with newPath.
// On Unix this renames over the existing file. If the target directory is
// not writable, it returns an error suggesting sudo.
func replaceBinary(oldPath, newPath string) error {
	dir := filepath.Dir(oldPath)

	// Check write permission.
	testFile := filepath.Join(dir, ".yishan-update-test")
	f, err := os.Create(testFile)
	if err != nil {
		return fmt.Errorf("no write permission to %s (try running with sudo)", dir)
	}
	f.Close()
	os.Remove(testFile)

	// Backup current binary.
	backupPath := oldPath + ".bak"
	if err := os.Rename(oldPath, backupPath); err != nil {
		return fmt.Errorf("backing up current binary: %w", err)
	}

	// Move new binary into place.
	if err := copyFile(newPath, oldPath); err != nil {
		// Restore backup on failure.
		_ = os.Rename(backupPath, oldPath)
		return fmt.Errorf("installing new binary: %w", err)
	}

	// Make executable.
	if err := os.Chmod(oldPath, 0755); err != nil {
		return fmt.Errorf("setting permissions: %w", err)
	}

	// Remove backup.
	os.Remove(backupPath)

	return nil
}

// copyFile copies src to dst across filesystems.
func copyFile(src, dst string) error {
	in, err := os.Open(src)
	if err != nil {
		return err
	}
	defer in.Close()

	out, err := os.OpenFile(dst, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0755)
	if err != nil {
		return err
	}
	defer out.Close()

	_, err = io.Copy(out, in)
	return err
}

// isHomebrewManaged returns true if the binary path is inside a Homebrew Cellar.
func isHomebrewManaged(execPath string) bool {
	// Homebrew Cellar paths:
	//   /opt/homebrew/Cellar/yishan/... (Apple Silicon)
	//   /usr/local/Cellar/yishan/...    (Intel Mac)
	//   /home/linuxbrew/.linuxbrew/Cellar/... (Linux)
	return strings.Contains(execPath, "/Cellar/")
}

func applyViaHomebrew(_ context.Context, progress func(string)) error {
	brewPath, err := exec.LookPath("brew")
	if err != nil {
		return fmt.Errorf("binary is Homebrew-managed but brew not found: %w", err)
	}

	progress("Updating Homebrew tap...")
	tapCmd := exec.Command(brewPath, "tap", "yishan-io/tap")
	tapCmd.Stdout = os.Stdout
	tapCmd.Stderr = os.Stderr
	// Ignore tap errors — may already be tapped.
	_ = tapCmd.Run()

	progress("Upgrading via Homebrew...")
	upgradeCmd := exec.Command(brewPath, "upgrade", "yishan")
	upgradeCmd.Stdout = os.Stdout
	upgradeCmd.Stderr = os.Stderr
	if err := upgradeCmd.Run(); err != nil {
		// brew upgrade exits non-zero if already up to date; try reinstall.
		reinstallCmd := exec.Command(brewPath, "reinstall", "yishan")
		reinstallCmd.Stdout = os.Stdout
		reinstallCmd.Stderr = os.Stderr
		if reinstallErr := reinstallCmd.Run(); reinstallErr != nil {
			return fmt.Errorf("brew upgrade failed: %w", err)
		}
	}

	return nil
}

func normalizeOS(goos string) string {
	return goos // "darwin", "linux", "windows" — already correct.
}

func normalizeArch(goarch string) string {
	return goarch // "amd64", "arm64" — already correct.
}

File: apps/cli-rs/src/cmd/self_update.rs
/// Self-update command — downloads, verifies, and replaces the current binary.
///
/// For Homebrew-managed installs the update is delegated to `brew upgrade`.
use crate::buildinfo::VERSION;
use crate::runtime::AppRuntime;
use clap::Args;
use sha2::{Digest, Sha256};
use std::env;
use std::fs;
use std::io::{self, Read};
use std::path::{Path, PathBuf};

const REPO: &str = "yishan-io/yishan-mono";
const PROJECT: &str = "yishan-cli";
const BINARY_NAME: &str = "yishan";
const RELEASE_TAG_PREFIX: &str = "cli-v";

#[derive(Args)]
pub struct SelfUpdateArgs {
    /// Target version (default: latest)
    pub version: Option<String>,

    /// Re-install even if already up to date
    #[arg(short, long)]
    pub force: bool,
}

pub async fn run(args: SelfUpdateArgs, _runtime: &AppRuntime) -> anyhow::Result<()> {
    let release = if let Some(v) = &args.version {
        println!("Target version: {v}");
        release_for_version(v)
    } else {
        println!("Checking for updates...");
        let r = latest_release().await?;
        println!("Latest version: {}", r.version);
        r
    };

    println!("Current version: {VERSION}");

    if !args.force && !is_newer(&release.version) {
        println!("Already up to date.");
        return Ok(());
    }

    release.apply(|msg| println!("{msg}")).await
}

// ── Release descriptor ────────────────────────────────────────────────────────

struct Release {
    version: String,
    archive_url: String,
    checksum_url: String,
}

fn release_for_version(version: &str) -> Release {
    let os_name = normalize_os(std::env::consts::OS);
    let arch = normalize_arch(std::env::consts::ARCH);
    let archive = format!("{PROJECT}_{version}_{os_name}_{arch}.tar.gz");
    let base = format!(
        "https://github.com/{REPO}/releases/download/{}",
        release_tag(version)
    );
    Release {
        version: version.to_string(),
        archive_url: format!("{base}/{archive}"),
        checksum_url: format!("{base}/checksums.txt"),
    }
}

async fn latest_release() -> anyhow::Result<Release> {
    // GitHub redirects /releases/latest → /releases/tag/<tag>.
    // We follow no redirects and extract the Location header.
    let url = format!("https://github.com/{REPO}/releases/latest");
    let client = reqwest::Client::builder()
        .redirect(reqwest::redirect::Policy::none())
        .build()?;
    let resp = client.head(&url).send().await?;
    let location = resp
        .headers()
        .get(reqwest::header::LOCATION)
        .and_then(|v| v.to_str().ok())
        .ok_or_else(|| {
            anyhow::anyhow!(
                "no redirect from GitHub releases/latest (status {})",
                resp.status()
            )
        })?;

    // Location: https://github.com/yishan-io/yishan-mono/releases/tag/cli-v0.12.0
    let tag = location
        .split("/tag/")
        .nth(1)
        .filter(|s| !s.is_empty())
        .ok_or_else(|| anyhow::anyhow!("unexpected release tag format: {location}"))?;
    let version = parse_release_version(tag)
        .ok_or_else(|| anyhow::anyhow!("unsupported release tag format: {tag}"))?;

    Ok(release_for_version(version))
}

fn release_tag(version: &str) -> String {
    format!("{RELEASE_TAG_PREFIX}{version}")
}

fn parse_release_version(tag: &str) -> Option<&str> {
    tag.strip_prefix(RELEASE_TAG_PREFIX)
        .or_else(|| tag.strip_prefix('v'))
}

fn is_newer(release_version: &str) -> bool {
    release_version != VERSION && VERSION != "dev"
}

impl Release {
    async fn apply(&self, progress: impl Fn(&str)) -> anyhow::Result<()> {
        // Resolve and canonicalise the running binary path.
        let exec_path = env::current_exe()?;
        let exec_path = fs::canonicalize(&exec_path)?;

        // Detect Homebrew-managed install.
        if is_homebrew_managed(&exec_path) {
            return apply_via_homebrew(&progress);
        }

        let tmp_dir = tempfile::tempdir()?;

        // Download archive.
        let archive_name = self
            .archive_url
            .split('/')
            .last()
            .unwrap_or("archive.tar.gz");
        progress(&format!("Downloading {archive_name}..."));
        let archive_path = tmp_dir.path().join("archive.tar.gz");
        download(&self.archive_url, &archive_path).await?;

        // Download checksums.
        progress("Verifying checksum...");
        let checksums_path = tmp_dir.path().join("checksums.txt");
        download(&self.checksum_url, &checksums_path).await?;

        // Verify checksum.
        verify_checksum(&archive_path, &checksums_path)?;

        // Extract binary.
        progress("Extracting...");
        let new_binary_path = tmp_dir.path().join(BINARY_NAME);
        extract_binary(&archive_path, &new_binary_path)?;

        // Replace current binary atomically.
        progress("Installing...");
        replace_binary(&exec_path, &new_binary_path)?;

        Ok(())
    }
}

// ── helpers ───────────────────────────────────────────────────────────────────

async fn download(url: &str, dest: &Path) -> anyhow::Result<()> {
    let resp = reqwest::get(url).await?;
    if !resp.status().is_success() {
        anyhow::bail!("HTTP {} from {url}", resp.status());
    }
    let bytes = resp.bytes().await?;
    fs::write(dest, &bytes)?;
    Ok(())
}

fn verify_checksum(archive_path: &Path, checksums_path: &Path) -> anyhow::Result<()> {
    let checksums = fs::read_to_string(checksums_path)?;
    let os_name = normalize_os(std::env::consts::OS);
    let arch = normalize_arch(std::env::consts::ARCH);
    let suffix = format!("{os_name}_{arch}");

    let expected_hash = checksums
        .lines()
        .filter_map(|line| {
            let mut parts = line.split_whitespace();
            let hash = parts.next()?;
            let name = parts.next()?;
            if name.contains(&suffix) {
                Some(hash.to_string())
            } else {
                None
            }
        })
        .next()
        .ok_or_else(|| {
            anyhow::anyhow!("no checksum found for {os_name}/{arch} in checksums.txt")
        })?;

    let mut f = fs::File::open(archive_path)?;
    let mut hasher = Sha256::new();
    let mut buf = vec![0u8; 65536];
    loop {
        let n = f.read(&mut buf)?;
        if n == 0 {
            break;
        }
        hasher.update(&buf[..n]);
    }
    let actual_hash = hex::encode(hasher.finalize());

    if actual_hash != expected_hash {
        anyhow::bail!("checksum mismatch: expected {expected_hash}, got {actual_hash}");
    }
    Ok(())
}

fn extract_binary(archive_path: &Path, dest: &Path) -> anyhow::Result<()> {
    let f = fs::File::open(archive_path)?;
    let gz = flate2::read::GzDecoder::new(f);
    let mut archive = tar::Archive::new(gz);
    let exe_name = format!("{BINARY_NAME}.exe");

    for entry in archive.entries()? {
        let mut entry = entry?;
        let path = entry.path()?;
        let name = path
            .file_name()
            .and_then(|n| n.to_str())
            .unwrap_or("")
            .to_string();
        if name == BINARY_NAME || name == exe_name {
            let mut out = fs::OpenOptions::new()
                .create(true)
                .write(true)
                .truncate(true)
                .open(dest)?;
            io::copy(&mut entry, &mut out)?;
            #[cfg(unix)]
            {
                use std::os::unix::fs::PermissionsExt;
                fs::set_permissions(dest, fs::Permissions::from_mode(0o755))?;
            }
            return Ok(());
        }
    }
    anyhow::bail!("binary {BINARY_NAME:?} not found in archive")
}

fn replace_binary(old_path: &Path, new_path: &Path) -> anyhow::Result<()> {
    let dir = old_path.parent().unwrap_or(Path::new("."));

    // Check write permission.
    let test_file = dir.join(".yishan-update-test");
    fs::File::create(&test_file).map_err(|_| {
        anyhow::anyhow!(
            "no write permission to {} (try running with sudo)",
            dir.display()
        )
    })?;
    let _ = fs::remove_file(&test_file);

    // Backup current binary.
    let backup_path = PathBuf::from(format!("{}.bak", old_path.display()));
    fs::rename(old_path, &backup_path)
        .map_err(|e| anyhow::anyhow!("backing up current binary: {e}"))?;

    // Copy new binary into place.
    if let Err(e) = fs::copy(new_path, old_path) {
        // Restore backup on failure.
        let _ = fs::rename(&backup_path, old_path);
        anyhow::bail!("installing new binary: {e}");
    }

    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        fs::set_permissions(old_path, fs::Permissions::from_mode(0o755))
            .map_err(|e| anyhow::anyhow!("setting permissions: {e}"))?;
    }

    let _ = fs::remove_file(&backup_path);
    Ok(())
}

fn is_homebrew_managed(exec_path: &Path) -> bool {
    exec_path.to_string_lossy().contains("/Cellar/")
}

fn apply_via_homebrew(progress: &impl Fn(&str)) -> anyhow::Result<()> {
    let brew = which::which("brew")
        .map_err(|_| anyhow::anyhow!("binary is Homebrew-managed but brew not found in PATH"))?;

    progress("Updating Homebrew tap...");
    // Ignore tap errors — may already be tapped.
    let _ = std::process::Command::new(&brew)
        .args(["tap", "yishan-io/tap"])
        .status();

    progress("Upgrading via Homebrew...");
    let status = std::process::Command::new(&brew)
        .args(["upgrade", BINARY_NAME])
        .status()?;
    if !status.success() {
        // brew upgrade exits non-zero if already up to date — try reinstall.
        let status = std::process::Command::new(&brew)
            .args(["reinstall", BINARY_NAME])
            .status()?;
        if !status.success() {
            anyhow::bail!("brew upgrade/reinstall failed");
        }
    }
    Ok(())
}

fn normalize_os(os: &str) -> &'static str {
    match os {
        "macos" => "darwin",
        other => Box::leak(other.to_string().into_boxed_str()),
    }
}

fn normalize_arch(arch: &str) -> &'static str {
    match arch {
        "x86_64" => "amd64",
        "aarch64" => "arm64",
        other => Box::leak(other.to_string().into_boxed_str()),
    }
}
